clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -111,12 +126,401 @@ class image_mem_wrapper {
+
+  sycl::ext::oneapi::experimental::interop_mem_handle
+      *_res_interop_mem_handle_ptr = nullptr;
+};
+
+#ifdef _WIN32
+class interop_mem_wrapper : public interop_mem_wrapper_base {
+public:
+  /// Creates interop memory wrapper for a given DX11 resource.
+  /// \param [in] d3d11_res Pointer to the ID3D11Resource to be wrapped.
+  /// \param [in] reg_flags Registration flags used for resource registration.
+  interop_mem_wrapper(ID3D11Resource *d3d11_res, unsigned reg_flags)
+      : _res_D3D11(d3d11_res), _res_reg_flags(reg_flags) {
+    // Check if the resource is already registered to prevent double
+    // registration.
+    if (_res_D3D11_set.find(_res_D3D11) != _res_D3D11_set.end()) {
+      throw std::runtime_error("This D3D11 resource is already registered!");
+    }
+
+    // Query the information from D3D11 resource object
+    query_res_info(_res_D3D11);
+
+    // Create a shared handle for the DX11 resource to enable interop.
+    _res_win_nt_handle = create_shared_handle(_res_D3D11);
+
+    // Prepare the external memory descriptor for SYCL interop.
+    sycl::ext::oneapi::experimental::resource_win32_handle ext_mem_win_handle{
+        _res_win_nt_handle};
+
+    sycl::ext::oneapi::experimental::external_mem_descriptor<
+        sycl::ext::oneapi::experimental::resource_win32_handle>
+        ext_mem_desc{ext_mem_win_handle,
+                     sycl::ext::oneapi::experimental::external_mem_handle_type::
+                         win32_nt_dx12_resource,
+                     _res_size_bytes};
+
+    // Import the external memory into SYCL for use with SYCL APIs.
+    new (_res_interop_mem_handle_ptr)
+        sycl::ext::oneapi::experimental::interop_mem_handle(
+            sycl::ext::oneapi::experimental::import_external_memory(
+                ext_mem_desc, get_default_queue()));
+
+    // Register the resource to track its uniqueness.
+    _res_D3D11_set.insert(_res_D3D11);
+  }
+
+  /// Cleans up the interop wrapper by releasing the SYCL interop
+  /// memory handle and the shared handle created for the DX11 resource.
+  virtual ~interop_mem_wrapper() {
+    if (!_res_interop_mem_handle_ptr) {
+      throw std::runtime_error(
+          "Resource is not registered! "
+          "Resource should be registered before unregistration.");
+    }
+
+    // Close the shared handle to the DX11 resource.
+    CloseHandle(_res_win_nt_handle);
+
+    // Release the external memory associated with the SYCL interop handle.
+    sycl::ext::oneapi::experimental::release_external_memory(
+        get_interop_mem_handle(), get_default_queue());
+
+    // Remove the resource from the tracking set.
+    _res_D3D11_set.erase(_res_D3D11);
+  }
+
+private:
+  static std::set<ID3D11Resource *> _res_D3D11_set;
+
+  ID3D11Resource *_res_D3D11 = nullptr;
+  HANDLE _res_win_nt_handle = nullptr;
+  unsigned _res_reg_flags = 0;
+  sycl::ext::oneapi::experimental::interop_mem_handle
+      *_res_interop_mem_handle_ptr = nullptr;
```

Comment:
Why have this member in both derived and base classes?

Reply:
It was a faulty commit. I've corrected it now.